clippy_lints/src/static_mut.rs

```diff
@@ -0,0 +1,60 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use rustc_ast::ast::{Item, ItemKind, Mutability, StaticItem};
+use rustc_lint::{EarlyContext, EarlyLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::declare_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Produces warnings when a `static mut` is declared.
+    ///
+    /// ### Why is this bad?
+    /// `static mut` can [easily produce undefined behavior][1] and
+    /// [may be removed in the future][2].
+    ///
+    /// ### Example
+    /// ```no_run
+    /// static mut GLOBAL_INT: u8 = 0;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// use std::sync::RwLock;
+    ///
+    /// static GLOBAL_INT: RwLock<u8> = RwLock::new(0);
```

Comment:
I think we should extend the example to also show at least one of the `Atomic*` types, as well as possibly `Lazy` or perhaps even `UnsafeCell`.

Reply:
correctly using `Atomic*` is tricky, and i don't think we should be reccomending them to beginners.  `UnsafeCell` won't work, we need `SyncUnsafeCell`, which isn't stable yet.

`LazyLock` and `LazyCell` are both still unstable.

`OnceCell` is a decent option, however, it makes it pretty easy to introduce TOC-TOU bugs.

Reply:
I agree that all of these types have their pros and cons, yet just showing `RwLock` makes the docs woefully incomplete.

How about at least listing the options with a short description of their niche and shortcomings? Such as:

* If you just have an integer type and want to avoid blocking, use the appropriate sized `std::sync::atomic::Atomic*` type
* If you just need the mutation for setting up the data because the code isn't `const` and don't actually need to update later on, you can use `std::sync::Lazy`
* Or if you are in an application and can call the setup in `main` before starting any threads, you can use `std::cell::Once`
* If you can manage the unsafety, one of the `std::cell::*Cell` types might work for you, but be aware that the other options avoid the unsafety, often without measurably affecting performance.

Reply:
unless things have changed since i last checked, the type of a static must *always* implement `Sync`, which means most of these will *never* work.  i suppose i could mention the atomic primatives, and maybe `LazyLock` and `UnsafeSyncCell` under a "if you're already using nightly" section.

Reply:
`std::sync::Once` and the `std::sync::atomic`s do implement `Sync` last I checked. I'm on mobile now, so I won't check all of them, but that certainly gives us some options.